### PR TITLE
Dev/databackendretake

### PR DESCRIPTION
**Pull request by LaurenSpiegel**
Implement a data file backend.

Functional tests pass.

This assumes that /data/pathNumbers exist on machine.  Do we want to create that through the Docker image and instruct in the readme or handle some other way?

### CODE REVIEW

**Comment by LaurenSpiegel** [comment]
Ready for final reviews.  Thanks, @RachedBenMustapha for all of the heavy lifting on the file flags!

I can drop the changes to the metadata wrapper depending on which PR gets merged first (this or the metadata file backend).

Since the ReadMe for master has diverged, I will update the readMe separately.

**Comment by LaurenSpiegel** [comment]
@ironman-machine any concerns?

File: lib/data/file/backend.js [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                log.error('error streaming data from request', { error: err });
+                fs.close(fd);
+                return callback(errors.InternalError);
+            })
+            .on('end', () => {
+                // eslint-disable-next-line no-param-reassign
+                request.calculatedHash = hash.digest('hex');
+                log.debug('finished writing data', { key,
+                    calculatedHash: request.calculatedHash });
+                fs.close(fd);
```
Comment (RachedBenMustapha) [inline lib/data/file/backend.js]: Here would be a great place to issue a please-really-write-to-disk request: https://nodejs.org/docs/v4.4.4/api/fs.html#fs_fs_fsync_fd_callback

It still does not guarantee 100% that all is written to the physical device, but it gets us closer.

File: lib/data/file/backend.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        }
+        const rs = fs.createReadStream(filePath, readStreamOptions);
+        return callback(null, rs);
+    },
+
+    delete: function delFile(key, reqUids, callback) {
+        const log = createLogger(reqUids);
+        const filePath = getFilePath(key);
+        log.debug('deleting file', { filePath });
+        return fs.unlink(filePath, callback);
```
Comment (RachedBenMustapha) [inline lib/data/file/backend.js]: `fs.unlink()` will give an `ENOENT` error to the callback so I think the error is not silently ignored. In any case, we should be aware if the metadata backend thinks an object still exists while it's missing on disk: it meands there is either a bug in some of our code or data/fs integrity issue - and we should be aware if that's the case.

File: init.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    const status = ioctl(pathFD, GETFLAGS, buffer);
+    assert.strictEqual(status, 0);
+    const currentFlags = buffer.readUIntLE(0, 8);
+    const flags = currentFlags | FS_DIRSYNC_FL;
+    buffer.writeUIntLE(flags, 0, 8);
+    const status2 = ioctl(pathFD, SETFLAGS, buffer);
+    assert.strictEqual(status2, 0);
+    const confirmBuffer = new Buffer(8);
+    ioctl(pathFD, GETFLAGS, confirmBuffer);
+    assert.strictEqual(confirmBuffer.readUIntLE(0, 8),
```
Comment (RachedBenMustapha) [inline init.js]: Not a big deal, but we could arguably want to do this assert on the value retrieved from a closed and re-opened fd.

File: init.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+} else {
+    logger.warn('WARNING: Synchronization directory updates are not ' +
+        'supported on this platform. Newly written data could be lost ' +
+        'if your system crashes before the operating system is able to ' +
+        'write directory updates.');
+}
+
+// Create 3511 subdirectories for the data file backend
+const subDirs = Array.from({ length: constants.folderHash },
+    (v, k) => (k + 1).toString());
```
Comment (RachedBenMustapha) [inline init.js]: Why aren't we zero-based?

File: init.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    fs.mkdir(`${dataPath}/${subDirName}`, err => {
+        // If already exists, move on
+        if (err && err.errno !== -17) {
+            return next(err);
+        }
+        return next();
+    });
+},
+ err => {
+     assert.strictEqual(err, null, `Error creating data files ${err}`);
```
Comment (RachedBenMustapha) [inline init.js]: Does `assert` work as expected (interrupting the whole program flow) when called from several layers of callbacks?

File: lib/data/file/backend.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+
+function getFilePath(key) {
+    const hash = stringHash(key);
+    const folderHashPath = ((hash % FOLDER_HASH) + 1).toString();
+    return `${STORAGE_PATH}/${folderHashPath}/${key}`;
+}
+
+export const backend = {
+    put: function putFile(request, keyContext, reqUids, callback) {
```
Comment (RachedBenMustapha) [inline lib/data/file/backend.js]: I'm not a fan of leaking the HTTP request here but I guess it's not specific to this file backend. Would it work like a `ReadableStream` if we just call it `stream` or do we actually use features from the request?

**Comment by RachedBenMustapha** [comment]
A few minor comments/questions, but I'm 👍 .

File: lib/data/file/backend.js (hunk 6 above)
Comment (LaurenSpiegel) [inline lib/data/file/backend.js]: We add the calculatedHash to it like with the real data backend.  Should be cleaned up in a separate PR.

File: init.js (hunk 5 above)
Comment (LaurenSpiegel) [inline init.js]: Yes.  Just confirmed by testing.

**Comment by vrancurel** [comment]
let's use localData and localMetadata names.

File: lib/data/file/backend.js [hunk 7]
```diff
@@ -1,93 +1,110 @@
 import { errors } from 'arsenal';
 import crypto from 'crypto';
-import stream from 'stream';
 import fs from 'fs';
+import config from '../../Config';
+import Logger from 'werelogs';
```
Comment (vrancurel) [inline lib/data/file/backend.js]: Please use in order:
core dependencies
nl
external dependencies
nl
internal dependencies

File: lib/data/file/backend.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            .on('open', () => { callback(null, rs); });
+    },
+
+    delete: function delFile(key, reqUids, callback) {
+        const log = createLogger(reqUids);
+        const filePath = getFilePath(key);
+        log.debug('deleting file', { filePath });
+        return fs.unlink(filePath, err => {
+            if (err) {
+                log.error('error deleting file', { error: err });
```
Comment (RachedBenMustapha) [inline lib/data/file/backend.js]: We could maybe include the filename in the log attributes for ops people to easily look at the issue?

File: constants.js [hunk 9]
```diff
@@ -49,4 +49,7 @@ export default {
     // ACLs.
     logId: 'http://acs.amazonaws.com/groups/s3/LogDelivery',
     emptyFileMd5: 'd41d8cd98f00b204e9800998ecf8427e',
+
+    // Number of sub-directories for file backend
+    folderHash: 3511, // Prime number
```
Comment (RachedBenMustapha) [inline constants.js]: Did we run tests to see which prime number yields better results depending on the filesystem type? Just kidding 😆

Comment (LaurenSpiegel) [inline constants.js]: That's @GiorgioRegni 's hobby.

**Comment by vrancurel** [comment]
👍

**Comment by LaurenSpiegel** [comment]
@ironman-machine SVC_BASE_FORCE_BUILD=1
NODESVC_BASE_FORCE_BUILD=1

**Comment by LaurenSpiegel** [comment]
@ironman-machine SVC_BASE_FORCE_BUILD=1
NODESVC_BASE_FORCE_BUILD=1

**Comment by LaurenSpiegel** [comment]
@ironman-machine where are you
SVC_BASE_FORCE_BUILD=1
NODESVC_BASE_FORCE_BUILD=1

**Comment by rahulreddy** [comment]
👍

**Comment by LaurenSpiegel** [comment]
@ironman-machine get to it

**Comment by DavidPineauScality** [comment]
@ironman-machine No sleep for the warriors, go and charge !

**Comment by DavidPineauScality** [comment]
Merging on behalf of @LaurenSpiegel since End-to-End passes :) Congrats ;)
